# What causes antibiotics resistance?
*Мельникова Дана, Чеченина Анна*

Ниже перечисленны все команды и результаты, которые мы использовали в своей работе. Все комманды были продублированы локально и финальные результаты, приведенные в отчете, соответствуют прогону на **Ubuntu 18.04.6 LTS bionic**.

## Установка FastQC

In [ ]:
!java -version

openjdk version "11.0.16" 2022-07-19
OpenJDK Runtime Environment (build 11.0.16+8-post-Ubuntu-0ubuntu118.04)
OpenJDK 64-Bit Server VM (build 11.0.16+8-post-Ubuntu-0ubuntu118.04, mixed mode, sharing)


In [ ]:
!wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip

--2022-11-02 23:29:57--  https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip
Resolving www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)... 149.155.133.4
Connecting to www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)|149.155.133.4|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10249221 (9.8M) [application/zip]
Saving to: ‘fastqc_v0.11.9.zip’

fastqc_v0.11.9.zip  100%[===================>]   9.77M  1.70MB/s    in 6.2s    

2022-11-02 23:30:04 (1.59 MB/s) - ‘fastqc_v0.11.9.zip’ saved [10249221/10249221]



In [ ]:
!unzip fastqc_v0.11.9.zip

Archive:  fastqc_v0.11.9.zip
  inflating: FastQC/cisd-jhdf5.jar   
   creating: FastQC/Configuration/
  inflating: FastQC/Configuration/adapter_list.txt  
  inflating: FastQC/Configuration/contaminant_list.txt  
  inflating: FastQC/Configuration/limits.txt  
  inflating: FastQC/fastqc           
  inflating: FastQC/fastqc_icon.ico  
   creating: FastQC/Help/
   creating: FastQC/Help/1 Introduction/
   creating: FastQC/Help/1 Introduction/.svn/
  inflating: FastQC/Help/1 Introduction/.svn/entries  
   creating: FastQC/Help/1 Introduction/.svn/props/
   creating: FastQC/Help/1 Introduction/.svn/text-base/
  inflating: FastQC/Help/1 Introduction/.svn/text-base/1.1 What is FastQC.html.svn-base  
   creating: FastQC/Help/1 Introduction/.svn/tmp/
   creating: FastQC/Help/1 Introduction/.svn/tmp/props/
  inflating: FastQC/Help/1 Introduction/1.1 What is FastQC.html  
   creating: FastQC/Help/2 Basic Operations/
   creating: FastQC/Help/2 Basic Operations/.svn/
  inflating: FastQC/Help/2 Basic

In [ ]:
!chmod a+x FastQC/fastqc

In [ ]:
!./FastQC/fastqc --help


            FastQC - A high throughput sequence QC analysis tool

SYNOPSIS

	fastqc seqfile1 seqfile2 .. seqfileN

    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] 
           [-c contaminant file] seqfile1 .. seqfileN

DESCRIPTION

    FastQC reads a set of sequence files and produces from each one a quality
    control report consisting of a number of different modules, each one of 
    which will help to identify a different potential type of problem in your
    data.
    
    If no files to process are specified on the command line then the program
    will start as an interactive graphical application.  If files are provided
    on the command line then the program will run with no user interaction
    required.  In this mode it is suitable for inclusion into a standardised
    analysis pipeline.
    
    The options for the program as as follows:
    
    -h --help       Print this help file and exit
    
    -v --version    Print the version of the program and exit

## Установка других необходимых пакетов



In [ ]:
pip install -q condacolab

In [ ]:
import condacolab

In [ ]:
condacolab.install()

⏬ Downloading https://github.com/jaimergp/miniforge/releases/latest/download/Mambaforge-colab-Linux-x86_64.sh...
📦 Installing...
📌 Adjusting configuration...
🩹 Patching environment...
⏲ Done in 0:00:21
🔁 Restarting kernel...


In [ ]:
!conda install -c bioconda trimmomatic

Solving environment: | / - \ | / - \ | / - \ | / done

## Package Plan ##

  environment location: /usr/local

  added / updated specs:
    - trimmomatic


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2022.9.24  |       ha878542_0         150 KB  conda-forge
    certifi-2022.9.24          |     pyhd8ed1ab_0         155 KB  conda-forge
    conda-22.9.0               |   py37h89c1867_1         960 KB  conda-forge
    openjdk-8.0.332            |       h166bdaf_0        97.8 MB  conda-forge
    openssl-1.1.1s             |       h166bdaf_0         2.1 MB  conda-forge
    trimmomatic-0.39           |       hdfd78af_2         144 KB  bioconda
    ------------------------------------------------------------
                                           Total:       101.3 MB

The following NEW packages will be INSTALLED:

  openjdk            conda-fo

In [ ]:
!apt-get install bowtie2 samtools

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  cwltool libc-ares2 libhts2 libjs-bootstrap nodejs nodejs-doc
  python-asn1crypto python-avro python-cachecontrol python-certifi
  python-cffi-backend python-chardet python-cryptography python-enum34
  python-html5lib python-idna python-ipaddress python-isodate python-lockfile
  python-mistune python-openssl python-pkg-resources python-pyparsing
  python-rdflib python-rdflib-jsonld python-requests python-ruamel.yaml
  python-schema-salad python-shellescape python-six python-sparqlwrapper
  python-typing python-urllib3 python-webencodings
Suggested packages:
  bowtie2-examples python-cryptography-doc python-cryptography-vectors
  python-enum34-doc python-genshi python-lxml python-lockfile-doc
  python-opens

## Загрузка данных и запуск FastQC

In [ ]:
# resistant
!wget 'https://figshare.com/ndownloader/files/23769689' -O amp_res_1.fastq.gz
!wget 'https://figshare.com/ndownloader/files/23769692' -O amp_res_2.fastq.gz
# non-resistant
!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz
!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gff.gz

--2022-11-02 23:31:42--  https://figshare.com/ndownloader/files/23769689
Resolving figshare.com (figshare.com)... 34.249.11.40, 34.248.76.93, 2a05:d018:1f4:d000:e0da:3116:1b67:a60b, ...
Connecting to figshare.com (figshare.com)|34.249.11.40|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769689/amp_res_1.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20221102/eu-west-1/s3/aws4_request&X-Amz-Date=20221102T233142Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=ccecfb2f838504c60aa236810b00e449cdca11bec47c72a259aa5b50c9d67f0d [following]
--2022-11-02 23:31:42--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769689/amp_res_1.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20221102/eu-west-1/s3/aws4_request&X-Amz-Date=20221102T233142Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=ccecfb2f838504c60aa236810b00e449

In [ ]:
# распаковка ахрива
!gzip -d /content/GCF_000005845.2_ASM584v2_genomic.fna.gz /content/GCF_000005845.2_ASM584v2_genomic.gff.gz /content/amp_res_1.fastq.gz /content/amp_res_2.fastq.gz

In [ ]:
!head -20 amp_res_1.fastq

@SRR1363257.37 GWZHISEQ01:153:C1W31ACXX:5:1101:14027:2198 length=101
GGTTGCAGATTCGCAGTGTCGCTGTTCCAGCGCATCACATCTTTGATGTTCACGCCGTGGCGTTTAGCAATGCTTGAAAGCGAATCGCCTTTGCCCACACG
+
@?:=:;DBFADH;CAECEE@@E:FFHGAE4?C?DE<BFGEC>?>FHE4BFFIIFHIBABEECA83;>>@>@CCCDC9@@CC08<@?@BB@9:CC#######
@SRR1363257.46 GWZHISEQ01:153:C1W31ACXX:5:1101:19721:2155 length=101
GTATGAGGTTTTGCTGCATTCTCTGNGCGAATATTAACTCCNTNNNNNTTATAGTTCAAAGCAAGTACCTGTCTCTTATACACATCTCCGAGCCCACGAGC
+
@@<?=D?D==?<AFGDF+AIHEACH#22<:?E8??:9??GG#0#####000;CF=C)4.==CA@@@)=7?C7?E37;3@>;;(.;>AB#############
@SRR1363257.77 GWZHISEQ01:153:C1W31ACXX:5:1101:5069:2307 length=101
GCTTCTCTTAACTGAGGTCACCATCATGCCGTTAAGTCCCTACCTCTCTTTTGCCGGTAACTGTTCCGCCGCGATTGCCTTTTATCTGTCTCTTATACACC
+
??<DBD;4C2=<BB>:AC;<CF<CE@FE9@E1C@891CD*9:?:3D@DD4?D<DD:0;@A=AEIDDA##################################
@SRR1363257.78 GWZHISEQ01:153:C1W31ACXX:5:1101:5178:2440 length=101
GCATAAGGACGATCGCTCCAGAGTAAAATAAATACGCGCATGTGATACTCACAATACCAATGGTGAAGTTACGGGACTTAAACAAACTGAGATCAAGAATC
+
CCCF

In [ ]:
!head -20 amp_res_2.fastq

@SRR1363257.37 GWZHISEQ01:153:C1W31ACXX:5:1101:14027:2198 length=101
GATCTAAGCTGAAGCCAGGCCAAAGTTTGACGATTGGTGCAGGCAGTAGCGCACAGCGACTGGCAAACAACAGCGATAGCATTACGTATCGTGTGCGCAAA
+
???BDB:DFHBFD@9;;+A;AFGH;ABHFHHGE@9:B:??@D>@;F?D8<<F8AA9EHHD8'..;5?A?A992(',(59CC3@C>22::A238+2>B<>B<
@SRR1363257.46 GWZHISEQ01:153:C1W31ACXX:5:1101:19721:2155 length=101
GTACTTGCTTTGNACTATAATATGCACGGAGNTAATATTCGCTCAGAGAATGCAGCAAAACCTCATACCTGTCTCTTATACACATCTGACGCTGCCGACGA
+
;@@DB?B;CFBB#2<:CB:FH<C@:<A?C::#1:86:BG9:8?8688?888EBF;783)=6-7=CC;ECD);?7;;>>AE;>(5;->AC@;B@;8?#####
@SRR1363257.77 GWZHISEQ01:153:C1W31ACXX:5:1101:5069:2307 length=101
ATAATAGGCAATCGCGTCGGAACAGTTACCGGCCAAAGAGAGGCAGGGACTTAACGGCATGATGGTGACCTCAGTTAAGAGAAGCCTGTCTCTTATACACA
+
+=?;:2,+A++AC:C:2@F6:CD:B09B?4)8@''8=))8=;=((5=4@?;@6;@?@BB;(535::>:>3(::(44:@::@3((9<32+::@(4@4+:>C3
@SRR1363257.78 GWZHISEQ01:153:C1W31ACXX:5:1101:5178:2440 length=101
ATATTAACAGTAGTATCAGTTATTTCTCTGATCTCTTTAGTCATTTGGGAGTCGACCTCAGAGAACCCGATTCTTGATCTCAGTTTGTTTAAGTCCCGTAA
+
BCCF

In [ ]:
!head -20 GCF_000005845.2_ASM584v2_genomic.fna

>lcl|NC_000913.3_cds_NP_414542.1_1 [gene=thrL] [locus_tag=b0001] [db_xref=UniProtKB/Swiss-Prot:P0AD86] [protein=thr operon leader peptide] [protein_id=NP_414542.1] [location=190..255] [gbkey=CDS]
ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCACAGGTAACGGTGCGGGCTGA
>lcl|NC_000913.3_cds_NP_414543.1_2 [gene=thrA] [locus_tag=b0002] [db_xref=UniProtKB/Swiss-Prot:P00561] [protein=fused aspartate kinase/homoserine dehydrogenase 1] [protein_id=NP_414543.1] [location=337..2799] [gbkey=CDS]
ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAACGTTTTCTGCGTGTTGCCGATATTCTGGAAAGCAA
TGCCAGGCAGGGGCAGGTGGCCACCGTCCTCTCTGCCCCCGCCAAAATCACCAACCACCTGGTGGCGATGATTGAAAAAA
CCATTAGCGGCCAGGATGCTTTACCCAATATCAGCGATGCCGAACGTATTTTTGCCGAACTTTTGACGGGACTCGCCGCC
GCCCAGCCGGGGTTCCCGCTGGCGCAATTGAAAACTTTCGTCGATCAGGAATTTGCCCAAATAAAACATGTCCTGCATGG
CATTAGTTTGTTGGGGCAGTGCCCGGATAGCATCAACGCTGCGCTGATTTGCCGTGGCGAGAAAATGTCGATCGCCATTA
TGGCCGGCGTATTAGAAGCGCGCGGTCACAACGTTACTGTTATCGATCCGGTCGAAAAACTGCTGGCAGTGGGGCATTAC
CTCGAATCTACCGTCGATATTGCTGAGTCC

In [ ]:
!wc -l amp_res_1.fastq

1823504 amp_res_1.fastq


In [ ]:
!wc -l amp_res_2.fastq

1823504 amp_res_2.fastq


In [ ]:
1823504/4

455876.0

In [ ]:
#!apt-get install fastqc

In [ ]:
#!fastqc -h

In [ ]:
!./FastQC/fastqc amp_res_1.fastq
!./FastQC/fastqc amp_res_2.fastq

Started analysis of amp_res_1.fastq
Approx 5% complete for amp_res_1.fastq
Approx 10% complete for amp_res_1.fastq
Approx 15% complete for amp_res_1.fastq
Approx 20% complete for amp_res_1.fastq
Approx 25% complete for amp_res_1.fastq
Approx 30% complete for amp_res_1.fastq
Approx 35% complete for amp_res_1.fastq
Approx 40% complete for amp_res_1.fastq
Approx 45% complete for amp_res_1.fastq
Approx 50% complete for amp_res_1.fastq
Approx 55% complete for amp_res_1.fastq
Approx 60% complete for amp_res_1.fastq
Approx 65% complete for amp_res_1.fastq
Approx 70% complete for amp_res_1.fastq
Approx 75% complete for amp_res_1.fastq
Approx 80% complete for amp_res_1.fastq
Approx 85% complete for amp_res_1.fastq
Approx 90% complete for amp_res_1.fastq
Approx 95% complete for amp_res_1.fastq
Analysis complete for amp_res_1.fastq
Started analysis of amp_res_2.fastq
Approx 5% complete for amp_res_2.fastq
Approx 10% complete for amp_res_2.fastq
Approx 15% complete for amp_res_2.fastq
Approx 20% c

Число ридов в итоговом файле fastqc равно числу ридов, подсчитанному выше (455876). \\
Мы можем видеть, что качество ридов значительно ухудшается начиная с ≈ 80 позиции. Среднее качество прочтения среди всех ридов равно ≈ 36; GC-состав в среднем равен 33%. Кроме того можно отметить, что качество прочтения первых позиций ридов достаточно низкое, а также что на концах ряда ридов детектируется адаптер Nextera transposase sequence. \\

* Для того, чтобы решить проблему плохого качества прочтения в самом начале обычно для секвенирования берут фрагменты, которые будут покрывать интересующий нас регион, захватывая небольшие последовательности перед и после него, чтобы на центральный регион приходилось больше центральных частей ридов. \\
* Проблема низкого качества ближе к концу ридов достаточно стандартная и решается путем уменьшения длины ридов и увеличением глубины прочтения.  \\
* Часто на концах остаются адапторы (что мы и наблюдаем на графике), поэтому мы можем без потери качества подрезать концы ридов, и убрать знакомые нам последовательности адапторов.


## Тримминг

In [ ]:
!apt-get install trimmomatic

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  libjbzip2-java
The following NEW packages will be installed:
  libjbzip2-java trimmomatic
0 upgraded, 2 newly installed, 0 to remove and 4 not upgraded.
Need to get 791 kB of archives.
After this operation, 859 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libjbzip2-java all 0.9.1-4 [42.6 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 trimmomatic all 0.36+dfsg-3 [749 kB]
Fetched 791 kB in 1s (627 kB/s)
Selecting previously unselected package libjbzip2-java.
(Reading database ... 125405 files and directories currently installed.)
Preparing to unpack .../libjbzip2-java_0.9.1-4_all.deb ...
Unpacking libjbzip2-java (0.9.1-4) ...
Select

Запускаем **pair-ended** run

In [ ]:
!trimmomatic PE -phred33 amp_res_1.fastq amp_res_2.fastq output_1_paired.fastq output_1_unpaired.fastq output_2_paired.fastq output_2_unpaired.fastqc ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:True \
LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20  

TrimmomaticPE: Started with arguments:
 -phred33 amp_res_1.fastq amp_res_2.fastq output_1_paired.fastq output_1_unpaired.fastq output_2_paired.fastq output_2_unpaired.fastqc ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:True LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20
Multiple cores found: Using 2 threads
java.io.FileNotFoundException: /content/TruSeq3-PE.fa (No such file or directory)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at org.usadellab.trimmomatic.fasta.FastaParser.parse(FastaParser.java:54)
	at org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.loadSequences(IlluminaClippingTrimmer.java:110)
	at org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.makeIlluminaClippingTrimmer(IlluminaClippingTrimmer.java:71)
	at org.usadellab.trimmomatic.trim.TrimmerFactory.makeTrimmer(TrimmerFactory.java:32)
	at org.usadellab.trimmomatic.Trimmomatic.createTrimme

Input Read Pairs: 455876 Both Surviving: 446259 (97.89%) Forward Only Surviving: 9216 (2.02%) Reverse Only Surviving: 273 (0.06%) Dropped: 128 (0.03%) \\

Эти числа совпадают с числами при применении команды *wc -l*: 1785036/4 = 446259

In [ ]:
!wc -l output_1_paired.fastq

1785036 output_1_paired.fastq


In [ ]:
1785036/4

446259.0

In [ ]:
#!fastqc -o . /content/amp_res_1_trimmed.fastq /content/amp_res_2_trimmed.fastq

In [ ]:
!./FastQC/fastqc output_1_paired.fastq
!./FastQC/fastqc output_2_paired.fastq

Started analysis of output_1_paired.fastq
Approx 5% complete for output_1_paired.fastq
Approx 10% complete for output_1_paired.fastq
Approx 15% complete for output_1_paired.fastq
Approx 20% complete for output_1_paired.fastq
Approx 25% complete for output_1_paired.fastq
Approx 30% complete for output_1_paired.fastq
Approx 35% complete for output_1_paired.fastq
Approx 40% complete for output_1_paired.fastq
Approx 45% complete for output_1_paired.fastq
Approx 50% complete for output_1_paired.fastq
Approx 55% complete for output_1_paired.fastq
Approx 60% complete for output_1_paired.fastq
Approx 65% complete for output_1_paired.fastq
Approx 70% complete for output_1_paired.fastq
Approx 75% complete for output_1_paired.fastq
Approx 80% complete for output_1_paired.fastq
Approx 85% complete for output_1_paired.fastq
Approx 90% complete for output_1_paired.fastq
Approx 95% complete for output_1_paired.fastq
Analysis complete for output_1_paired.fastq
Started analysis of output_2_paired.fastq

Мы можем видеть, что после тримминга параметр per base sequence quality заметно улучшился - вы видим, что квартили теперь уже и ближе к медиане и все находятся в "зеленой зоне" с качеством выше 28.

## Картирование ридов, получение sam и bam файлов

In [ ]:
!apt-get install bwa

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'apt autoremove' to remove it.
The following NEW packages will be installed:
  bwa
0 upgraded, 1 newly installed, 0 to remove and 4 not upgraded.
Need to get 176 kB of archives.
After this operation, 404 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 bwa amd64 0.7.17-1ubuntu0.1 [176 kB]
Fetched 176 kB in 1s (336 kB/s)
Selecting previously unselected package bwa.
(Reading database ... 125440 files and directories currently installed.)
Preparing to unpack .../bwa_0.7.17-1ubuntu0.1_amd64.deb ...
Unpacking bwa (0.7.17-1ubuntu0.1) ...
Setting up bwa (0.7.17-1ubuntu0.1) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...


In [ ]:
!bwa index


Usage:   bwa index [options] <in.fasta>

Options: -a STR    BWT construction algorithm: bwtsw, is or rb2 [auto]
         -p STR    prefix of the index [same as fasta name]
         -b INT    block size for the bwtsw algorithm (effective with -a bwtsw) [10000000]
         -6        index files named as <in.fasta>.64.* instead of <in.fasta>.* 

         `-a div' do not work not for long genomes.



In [ ]:
!bwa index GCF_000005845.2_ASM584v2_genomic.fna

[bwa_index] Pack FASTA... 0.02 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 0.70 seconds elapse.
[bwa_index] Update BWT... 0.02 sec
[bwa_index] Pack forward-only FASTA... 0.02 sec
[bwa_index] Construct SA from BWT and Occ... 0.29 sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa index GCF_000005845.2_ASM584v2_cds_from_genomic.fna
[main] Real time: 1.144 sec; CPU: 1.058 sec


In [ ]:
!bwa mem GCF_000005845.2_ASM584v2_genomic.fna output_1_paired.fastq output_2_paired.fastq > alignment.sam 

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 106336 sequences (10000178 bp)...
[M::process] read 108264 sequences (10000141 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 41216, 1, 11)
[M::mem_pestat] skip orientation FF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (137, 176, 221)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 389)
[M::mem_pestat] mean and std.dev: (180.19, 63.30)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 473)
[M::mem_pestat] skip orientation RF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation RR...
[M::mem_pestat] (25, 50, 75) percentile: (82, 117, 185)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 391)
[M::mem_pestat] mean and std.dev: (123.27, 60.10)
[M::mem_pestat] low and high boundaries for prop

In [ ]:
!samtools view -S -b alignment.sam > alignment.bam

In [ ]:
!samtools flagstat alignment.bam

901331 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
8813 + 0 supplementary
0 + 0 duplicates
798247 + 0 mapped (88.56% : N/A)
892518 + 0 paired in sequencing
446259 + 0 read1
446259 + 0 read2
715618 + 0 properly paired (80.18% : N/A)
759578 + 0 with itself and mate mapped
29856 + 0 singletons (3.35% : N/A)
42972 + 0 with mate mapped to a different chr
41407 + 0 with mate mapped to a different chr (mapQ>=5)


Overall, the 82.02% (88.56%) of reads are mapped.

*901331 + 0 in total (QC-passed reads + QC-failed reads)                                                                              0 + 0 secondary                                                                                                                      8813 + 0 supplementary                                                                                                               0 + 0 duplicates                                                                                                                     798247 + 0 mapped (88.56% : N/A)                                                                                                     892518 + 0 paired in sequencing                                                                                                      446259 + 0 read1                                                                                                                     446259 + 0 read2                                                                                                                     715618 + 0 properly paired (80.18% : N/A)                                                                                            759578 + 0 with itself and mate mapped                                                                                               29856 + 0 singletons (3.35% : N/A)                                                                                                   42972 + 0 with mate mapped to a different chr                                                                                        41407 + 0 with mate mapped to a different chr (mapQ>=5)*

In [ ]:
!samtools sort alignment.bam -o alignment_sorted.bam

In [ ]:
!samtools index alignment_sorted.bam

## Variant calling 

In [ ]:
!samtools mpileup -f GCF_000005845.2_ASM584v2_genomic.fna alignment_sorted.bam >  my.mpileup                            

[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000


Появился файл mpileup с сводкой по числу изменений по каждой позиции 

In [ ]:
!curl -L -k -o VarScan.v2.4.2.jar https://github.com/dkoboldt/varscan/releases/download/2.4.2/VarScan.v2.4.2.jar
!java -jar VarScan.v2.4.2.jar

curl: /usr/local/lib/libcurl.so.4: no version information available (required by curl)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  112k  100  112k    0     0   299k      0 --:--:-- --:--:-- --:--:--  299k
VarScan v2.4.2

***NON-COMMERCIAL VERSION***

USAGE: java -jar VarScan.jar [COMMAND] [OPTIONS] 

COMMANDS:
	pileup2snp		Identify SNPs from a pileup file
	pileup2indel		Identify indels a pileup file
	pileup2cns		Call consensus and variants from a pileup file
	mpileup2snp		Identify SNPs from an mpileup file
	mpileup2indel		Identify indels an mpileup file
	mpileup2cns		Call consensus and variants from an mpileup file

	somatic			Call germline/somatic variants from tumor-normal pileups
	copynumber			Determine relative tumor copy number from tumor-normal pileups
	readcounts		Obtain read counts f

In [ ]:
!java -jar VarScan.v2.4.2.jar mpileup2snp -h

Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.2
Min avg qual:	15
P-value thresh:	0.01
USAGE: java -jar VarScan.jar mpileup2cns [pileup file] OPTIONS
	mpileup file - The SAMtools mpileup file

	OPTIONS:
	--min-coverage	Minimum read depth at a position to make a call [8]
	--min-reads2	Minimum supporting reads at a position to call variants [2]
	--min-avg-qual	Minimum base quality at a position to count a read [15]
	--min-var-freq	Minimum variant allele frequency threshold [0.01]
	--min-freq-for-hom	Minimum frequency to call homozygote [0.75]
	--p-value	Default p-value threshold for calling variants [99e-02]
	--strand-filter	Ignore variants with >90% support on one strand [1]
	--output-vcf	If set to 1, outputs in VCF format
	--vcf-sample-list	For VCF output, a list of sample names in order, one per line
	--variants	Report only variant (SNP/indel) positions [0]


Only SNPs will be reported. Warning: No p-value threshold provided, so p-values will not be calculated. 
* Min coverage: 8
* Min reads2: 2
* Min var freq: 0.2
* Min avg qual: 15
* P-value thresh: 0.01 



Повысим требования к минимальному проценту оснований, отличающихся от референсных на каждой позиции, который требуется для идентификации мутации по этой позиции. По умолчанию это значение равно 0.01 (1%), возьмем 0.1 (10%).

In [ ]:
!java -jar VarScan.v2.4.2.jar  mpileup2snp my.mpileup --min-var-freq 0.1 --variants --output-vcf 1 > VarScan_results_10.vcf

* 9 variant positions (6 SNP, 3 indel)
* 1 were failed by the strand-filter
* 5 variant positions reported (5 SNP, 0 indel)

Мы видим, что при таком пороге программа обнаружила 6 SNP. Посмотрим, сколько оснований преодолели порог в 50%:


In [ ]:
!java -jar VarScan.v2.4.2.jar  mpileup2snp my.mpileup --min-var-freq 0.5 --variants --output-vcf 1 > VarScan_results_50.vcf

* 9 variant positions (6 SNP, 3 indel)
* 1 were failed by the strand-filter 
* 5 variant positions reported (5 SNP, 0 indel) 

Ничего не изменилось - нам остается только проаннотировать эти мутации!

In [ ]:
!mkdir -p data/k12
!cd ./data/k12

In [ ]:
!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gbff.gz

In [ ]:
!gunzip GCF_000005845.2_ASM584v2_genomic.gbff.gz 

In [ ]:
!nano snpEff.config # write k12.genome : ecoli_K12 there

In [ ]:
!cp GCF_000005845.2_ASM584v2_genomic.gbff data/k12/genes.gbk 
!cd ../../

In [ ]:
!java -jar snpEff.jar build -genbank -v k12

In [ ]:
!java -jar snpEff.jar ann k12 VarScan_results_50.vcf > VarScan_results_annotated.vcf 

## Genes


Printing nice table with frequencies:

In [12]:
!cat VarScan_results_annotated.vcf

##fileformat=VCFv4.1
##source=VarScan2
##INFO=<ID=ADP,Number=1,Type=Integer,Description="Average per-sample depth of bases with Phred score >= 15">
##INFO=<ID=WT,Number=1,Type=Integer,Description="Number of samples called reference (wild-type)">
##INFO=<ID=HET,Number=1,Type=Integer,Description="Number of samples called heterozygous-variant">
##INFO=<ID=HOM,Number=1,Type=Integer,Description="Number of samples called homozygous-variant">
##INFO=<ID=NC,Number=1,Type=Integer,Description="Number of samples not called">
##FILTER=<ID=str10,Description="Less than 10% or more than 90% of variant supporting reads on one strand">
##FILTER=<ID=indelError,Description="Likely artifact due to indel reads at this position">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
##FORMAT=<ID=SDP,Number=1,Type=Integer,Description="Raw Read Depth as reported by SAMtools">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Quality

In [19]:
!cat VarScan_results_annotated.vcf | awk 'NR>29 {print $1, $2, $4, $5, $8, $10}' > genes_vcf.txt

In [20]:
import pandas as pd

df = pd.read_csv("genes_vcf.txt", sep=" ", header=None, names=["Gene", "Position", "Before", "After", "Info", "Freq"])

In [21]:
df

Gene  Position Before After  \
0  NC_000913.3     93043      C     G   
1  NC_000913.3    482698      T     A   
2  NC_000913.3    852762      A     G   
3  NC_000913.3   3535147      A     C   
4  NC_000913.3   4390754      G     T   

                                                Info  \
0  ADP=17;WT=0;HET=0;HOM=1;NC=0;ANN=G|missense_va...   
1  ADP=16;WT=0;HET=0;HOM=1;NC=0;ANN=A|missense_va...   
2  ADP=14;WT=0;HET=0;HOM=1;NC=0;ANN=G|upstream_ge...   
3  ADP=18;WT=0;HET=0;HOM=1;NC=0;ANN=C|missense_va...   
4  ADP=15;WT=0;HET=0;HOM=1;NC=0;ANN=T|synonymous_...   

                                              Freq  
0  1/1:93:18:17:0:17:100%:4.2852E-10:0:36:0:0:6:11  
1   1/1:87:16:16:0:16:100%:1.6637E-9:0:45:0:0:12:4  
2    1/1:76:14:14:0:14:100%:2.4927E-8:0:36:0:0:9:5  
3  1/1:99:18:18:0:18:100%:1.1019E-10:0:34:0:0:12:6  
4    1/1:81:16:15:0:15:100%:6.4467E-9:0:36:0:0:8:7

In [23]:
df["Gene"] = df["Info"].apply(lambda x: x.split(";")[5][4:].split("|")[3])
df["Effect"] = df["Info"].apply(lambda x: x.split(";")[5][4:].split("|")[1])
df["Impact"] = df["Info"].apply(lambda x: x.split(";")[5][4:].split("|")[2])
df["GeneID"] = df["Info"].apply(lambda x: x.split(";")[5][4:].split("|")[4])
df["Freq"] = df["Freq"] = df["Freq"].apply(lambda x: float(x.split(":")[6][:-1]))
df.dtypes

Gene         object
Position      int64
Before       object
After        object
Info         object
Freq        float64
Effect       object
Impact       object
GeneID       object
dtype: object

In [25]:
df.drop(columns = ["Info"], inplace = True)

In [26]:
df.head()

Gene  Position Before After   Freq                 Effect    Impact GeneID
0  ftsI     93043      C     G  100.0       missense_variant  MODERATE  b0084
1  acrB    482698      T     A  100.0       missense_variant  MODERATE  b0462
2  glnH    852762      A     G  100.0  upstream_gene_variant  MODIFIER  b0811
3  envZ   3535147      A     C  100.0       missense_variant  MODERATE  b3404
4  rsgA   4390754      G     T  100.0     synonymous_variant       LOW  b4161